aiofile/utils.py

```diff
@@ -343,13 +349,52 @@ def async_open(
     return BinaryFileWrapper(afp)
 
 
+T = TypeVar("T", bound=FileIOWrapperBase)
+
+
+class FileIOCloner(Generic[T]):
+    def __init__(self, file: T):
+        self.source_afp = file
+        self.cloned_afp: Optional[T] = None
+        self.cloned_lock = asyncio.Lock()
+
+    async def __clone(self) -> T:
+        async with self.cloned_lock:
+            if self.cloned_afp is not None:
+                return self.cloned_afp
+
+            self.cloned_afp = self.source_afp.__class__(
+                await self.source_afp.file.clone(),
```

Comment:
What do you think about clone method in source-afp class? It would be more readable and easier to refactor in feature